### PR TITLE
add new JsonIgnoreAttribute WhenWriting/WhenReading

### LINKED ISSUE #66490: [API Proposal]: Extend `JsonIgnoreCondition`
### Background and motivation

Sometimes we may want to ignore some properties only when serialize, included when deserialize, currently, we had `JsonIgnoreCondition`

``` c#
public enum JsonIgnoreCondition
{
  Never,
  Always,
  WhenWritingDefault,
  WhenWritingNull,
}
```

Maybe we could add new conditions like `WhenWriting`/`WhenReading`(or `WhenSerializing`/`WhenDeserializing` ...)

### API Proposal

> edited

``` diff
namespace System.Text.Json.Serialization;

public enum JsonIgnoreCondition
{
  Never = 0,
  Always = 1,
  WhenWritingDefault = 2,
  WhenWritingNull = 3,
+ WhenWriting = 4,
+ WhenReading = 5,
}

```

### API Usage

```C#
public class TestModel
{  
  [JsonIgnore(Condition = JsonIgnoreCondition.WhenReading)]
  public int Age { get; set; }
  [JsonIgnore(Condition = JsonIgnoreCondition.WhenWriting)]
  public string Name { get; set; }
}

var json = JsonSerializer.Serialize(new TestModel{ Age = 10, Name="Mike" });
Assert.Equal("{\"Age\":10}", json);

json = "{\"Age\":10, \"Name\":\"Mike\"}";
var model = JsonSerializer.Deserialize<TestModel>(json);
Assert.Equal("Mike", model.Name);
Assert.Equal(0, model.Age);
```

### Alternative Designs

_No response_

### Risks

_No response_

### PR DESCRIPTION
**Pull request by WeihanLi**
feat: add new JsonIgnoreAttribute WhenWriting/WhenReading

fixes #66490 

- [x] test cases

### CODE REVIEW

File: src/libraries/System.Text.Json/Common/JsonIgnoreCondition.cs [hunk 1]
```diff
@@ -30,5 +30,15 @@ public enum JsonIgnoreCondition
...
+
+        /// <summary>
+        /// Property is ignored during serialization
+        /// </summary>
+        WhenWriting = 4,
+
+        /// <summary>
+        /// Property is ignored during deserialization
+        /// </summary>
+        WhenReading = 5,
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/Common/JsonIgnoreCondition.cs:42]: @eiriktsarpalis, how does someone specify multiple of these? e.g. if I have a property that I want to be ignored when reading and when writing null, how do I achieve that? Or is that not a scenario that matters?

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/Common/JsonIgnoreCondition.cs:42]: Not possible, which shows the limits of the `JsonIgnoreCondition` design. These enums eventually map to the `JsonPropertyInfo.ShouldSerialize` and `JsonPropertyInfo.Set` delegates, so a user wanting to introduce their own ignore conditions can configure these via the contract customization APIs.

**Comment by eiriktsarpalis** [comment]
@WeihanLi thanks for submitting the PR, however since we're currently in the process of finishing up .NET 9 we are no longer merging PRs adding new features. We can revisit this in a couple weeks' time once the .NET 9 branch has been snapped from main.

**Comment by eiriktsarpalis** [comment]
Apologies for the delay getting back to you, we can continue work on this PR for .NET 10 now.

**Review by eiriktsarpalis — APPROVED** [review]
Thanks!

**Pull request merged**